[DP-62] As a reader, I want a dedicated paragraph details pane

chanmioh opened the pull request:
### Changelog

- `segment.tsx`: Added a “details button” to the right of paragraphs, similar to the details button to the right of a select-able word.

## Syllabary View

![image](https://user-images.githubusercontent.com/40529597/188725698-c75b61e9-e0fe-4ea8-af10-b066f725f1ef.png)

## Syllabary, Simple Phonetics & Syllabary, Phonetics, and Word Parts View

![image](https://user-images.githubusercontent.com/40529597/188725641-2e47aecc-7167-404c-a711-ec06c479b8ef.png)

- `panel-layout.tsx`: Created new type of `FormFieldsFragment` or `TranslatedParagraph` to support selecting paragraphs.
    - `PanelLayout` returns a different layout depending on whether a word or paragraph is selected.
- `paragraph-panel.tsx`: Designed layout of paragraph panel display when a panel is selected.

![image](https://user-images.githubusercontent.com/40529597/188725748-8c5d7df9-4156-4cf6-8793-288cda86f7d4.png)

![image](https://user-images.githubusercontent.com/40529597/188725815-ef51a380-7921-41a5-877e-0199b65c1b08.png)

- `document.page.tsx`: Now supports selecting a word and a paragraph.
- `wordpress.tsx`: Modified to support selecting a word and a paragraph.

hazelyn11 commented on website/src/pages/documents/document.page.tsx:
```diff
@@ -128,28 +128,31 @@ const TranslationTab = ({ doc }: { doc: Document }) => {
   }
 
-  // When the mobile version of the word panel is closed, remove any word selection.
+  // When the mobile version of the word panel is closed, remove any segment selection.
```
Is segment here just referring to a paragraph? I know in other areas of the code, segment refers to a line and paragraphs are called blocks. I could also be misinterpreting what's going on here though

hazelyn11 replied:
Oh, actually, after looking a bit closer, it looks like now either a word or a paragraph could be considered a segment, which makes more sense!

chanmioh replied:
Yeah, I think throughout the code, the meaning of segment is pretty flexible. Glad to know it's not too confusing!

hazelyn11 commented on website/src/segment.tsx:
```diff
@@ -106,7 +142,9 @@ export const AnnotatedForm = (
-    const isSelected = p.wordPanelDetails.currContents?.id === p.segment.id
+    const isSelected =
+      (p.wordPanelDetails.currContents as Dailp.FormFieldsFragment)?.id ===
+      p.segment.id
```
Overall, I felt I could pretty easily follow your code despite the fact that I'm not familiar with this area of the project. It took me a second to realize that a segment refers to either a word or a paragraph, but you make this pretty clear with comments, and you're very consistent with naming things throughout.

hazelyn11 commented on website/src/pages/documents/document.page.tsx:
```diff
@@ -128,28 +128,31 @@ const TranslationTab = ({ doc }: { doc: Document }) => {
-  let wordPanelInfo = {
-    currContents: selectedWord,
-    setCurrContents: selectAndShowWord,
+  // This can now be either a word or paragraph.
```
Before merging, maybe just take "now" out of this comment since someone looking at the updated code won't see the old code

chanmioh replied:
Makes sense!

loafofpiecrust commented on website/src/paragraph-panel.tsx:
```diff
@@ -0,0 +1,57 @@
+        <h1 className={css.noSpaceBelow}>
+          Selected
+          <br />
+          Paragraph
```
Do you think we could show the paragraph number in the header? Same with the selected word. Instead of "selected word", "word #x"

loafofpiecrust reviewed (commented):
Do you think putting the selection icon centered to the right of the paragraph would be a good experience? Semantically where you put it makes sense since right now only the translations will be editable in the panel. But it does look perhaps a little awkward and easy to miss. Open to other ideas for placement.

chanmioh commented:
### Changelog
Modified some comments and the `DocumentParagraph` struct now contains an `index` field that represents a 1-indexed position of a paragraph within its document.

- `document.rs`: Added `index` field and its resolver
- `database_sql.rs`: Used enumerate() function to get the index of an object in an iterable, and assigned this as the index of a `DocumentParagraph`
- `queries.graphql`: Added `index` field to `DocumentContents` query
- `panel-layout.tsx` and `paragraph-panel.tsx`: Added index number to the panel header (see below)
<img width="362" alt="image" src="https://user-images.githubusercontent.com/40529597/193133758-f7836591-110d-420f-ae1a-210775120853.png">

<img width="370" alt="image" src="https://user-images.githubusercontent.com/40529597/193134051-5ba05381-889f-4150-bc3b-edcb6e1f9777.png">

- `edit-word-panel.tsx`: Removed `form.reset()` since this causes form to clear if user "cancels" but enters "edit" mode again.

chanmioh commented:
> Do you think putting the selection icon centered to the right of the paragraph would be a good experience? Semantically where you put it makes sense since right now only the translations will be editable in the panel. But it does look perhaps a little awkward and easy to miss. Open to other ideas for placement.

I'll have to give this some more thought - but I do agree the selection icon is a bit easy to miss at the moment.
